include/eos/fitting/ceres_nonlinear.hpp

```diff
@@ -420,31 +445,437 @@ std::array<T, 3> get_shape_point(const morphablemodel::PcaModel& shape_model,
+                                     &blendshape_coefficients[0], &colour_coefficients[0]);
+        }
+    }
+
+    /**
+     * Add residual block with shape prior cost function to the overall problem
+     *
+     * @param[in] weight constant to multiply all colours
+     * @param[in, out] upper and lower bound of each colour_coefficient
+     */
+    void add_image_prior_cost_function(double weigth = 35.0,
+                                       double color_coeff_limit = 3.0) {
+        // Templates: CostFunctor, num residuals, colour-coeffs
+        auto* colour_prior_cost =
+                new ceres::AutoDiffCostFunction<fitting::PriorCost, color_coeffs_num, color_coeffs_num>(
+                        new fitting::PriorCost(color_coeffs_num, weigth));
+
+        problem->AddResidualBlock(colour_prior_cost, nullptr, &colour_coefficients[0]);
+        for (int i = 0; i < color_coeffs_num; ++i) {
+            problem->SetParameterLowerBound(&colour_coefficients[0], i, -color_coeff_limit);
+            problem->SetParameterUpperBound(&colour_coefficients[0], i, color_coeff_limit);
+        }
+    }
+    #endif
+
+    /**
+     * Block shapes fitting
+     */
+    void block_shapes_fitting() {
+        problem->SetParameterBlockConstant(&shape_coefficients[0]);
+    }
+
+    /**
+     * Block blendshapes fitting
+     */
+    void block_blendshapes_fitting() {
+        problem->SetParameterBlockConstant(&blendshape_coefficients[0]);
+    }
+
+    /**
+     * Block fov fitting
+     *
+     * Note: it is only for perspective projection. It have no sense in case of ortogonal projection.
+     *
+     * @param[in,out] camera perspective camera to fit
+     * @param[in] fov_in_grad field of view to fix
+     */
+    void block_fov_fitting(PerspectiveCameraParameters& camera, double fov_in_grad) {
+        auto& camera_translation_and_intrinsics = camera.translation_and_intrinsics;
+        camera_translation_and_intrinsics[3] = glm::radians(fov_in_grad);
+
+        std::vector<int> vec_constant_extrinsic = {3};
+        auto subset_parameterization =
+                new ceres::SubsetParameterization(static_cast<int>(get_num_cam_params(true)),
+                                                  vec_constant_extrinsic);
+        problem->SetParameterization(&camera_translation_and_intrinsics[0], subset_parameterization);
+    }
+
+    darray<shapes_num> shape_coefficients;
+    darray<blendshapes_num> blendshape_coefficients;
+#if EOS_CERES_USE_OPENCV == true
+    darray<color_coeffs_num> colour_coefficients;
+#endif
+
+    std::unique_ptr<ceres::Problem> problem;
+private:
+    /*
+     * Check if blendshapes is not nullptr or try to find them in morphable_model.
+     *
+     * * @return std::array with 3 or 4 (for perspective projection) parameters.
+     */
+    auto find_blendshapes(const morphablemodel::Blendshapes* const blendshapes = nullptr) const {
+        if (blendshapes == nullptr) {
```

Comment:
I am not really sure I understand the purpose of this method. Why do we need a pointer that can be null, and could we not use `MorphableModel::has_separate_expression_model()` and `eos::cpp17::holds_alternative<...>(...)` for the variant?

Reply:
OK. I will change it.